File: PythonVersions/NumpyVersion/src/kd_tree.py
from __future__ import annotations

import numpy as np
from random import randrange
from dataclasses import dataclass
from numpy import sqrt
import numpy.typing as npt

from particle import calc_pp_accel, Particles


MAX_PARTS: int = 7
THETA: float = 0.3


@dataclass(slots=True)
class KDTree:
    # For leaves
    num_parts: int
    particles: list[int]

    # For internal nodes
    split_dim: int
    split_val: float
    m: float
    cm: tuple[float, float, float]  # de facto np.array
    size: float
    left: int
    right: int

    @staticmethod
    def leaf(num_parts: int, particles: list[int]) -> KDTree:
        return KDTree(
            num_parts,
            particles,
            None,  # type: ignore
            0.0,
            0.0,
            np.zeros(3, dtype=np.float64),
            0.0,
            None,  # type: ignore
            None  # type: ignore
        )


def allocate_node_vec(num_parts: int) -> list[KDTree]:
    num_nodes = 2 * (num_parts // (MAX_PARTS-1) + 1)
    return [KDTree.leaf(0, []) for _ in range(num_nodes)]


class System:
    indices: list[int]
    nodes: list[KDTree]

    __slots__ = ['indices', 'nodes']

    def __init__(self, n: int) -> None:
        self.indices = None
        self.nodes = allocate_node_vec(n)

    # Returns the index of the last Node used in the construction.
    def build_tree(self,
                   start: int,
                   end: int,
                   particles: Particles,
                   cur_node: int,
                   ) -> int:
        ps, vs, rs, ms = particles
        del particles

        # println!("start = {} end = {} cur_node = {}", start, end, cur_node)
        np1 = end - start
        # println!("s = {}, e = {}, cn = {}", start, end, cur_node)
        if np1 <= MAX_PARTS:
            if cur_node >= len(self.nodes):
                diff = cur_node + 1 - len(self.nodes)
                self.nodes.extend(KDTree.leaf(0, []) for _ in range(diff))
            self.nodes[cur_node].num_parts = np1

            self.nodes[cur_node].particles = [self.indices[start + i]
                                              for i in range(np1)]
            return cur_node
        else:
            # Pick split dim and value
            min = np.ones(3, dtype=np.float64) * 1e100
            max = np.ones(3, dtype=np.float64) * -1e100
            m = 0.0
            cm = np.zeros(3, dtype=np.float64)
            for i in range(start, end):
                m += ms[self.indices[i]]
                cm += ms[self.indices[i]] * \
                    ps[self.indices[i]]
                min = np.minimum(min, ps[self.indices[i]])
                max = np.maximum(max, ps[self.indices[i]])

            cm /= m
            split_dim = 0
            for dim in range(1, 2):  # FIXME: what is this
                if max[dim] - min[dim] > max[split_dim] - min[split_dim]:
                    split_dim = dim
            size = max[split_dim] - min[split_dim]

            # Partition particles on split_dim
            mid = (start + end) // 2
            s = start
            e = end
            while s + 1 < e:
                pivot = randrange(s, e)
                tmp = self.indices[s]
                try:
                    self.indices[s] = self.indices[pivot]
                except IndexError:
                    print(s, pivot, len(self.indices))
                self.indices[pivot] = tmp

                low = s + 1
                high = e - 1
                while low <= high:
                    if ps[self.indices[low]][split_dim] < ps[self.indices[s]][split_dim]:
                        low += 1
                    else:
                        tmp = self.indices[low]
                        self.indices[low] = self.indices[high]
                        self.indices[high] = tmp
                        high -= 1

                tmp = self.indices[s]
                self.indices[s] = self.indices[high]
                self.indices[high] = tmp

                if high < mid:
                    s = high + 1
                elif high > mid:
                    e = high
                else:
                    s = e

            split_val = ps[self.indices[mid]][split_dim]

            # Recurse on children and build this node.
            left = self.build_tree(start, mid,
                                   (ps, vs, rs, ms), cur_node + 1)
            right = self.build_tree(mid, end,
                                    (ps, vs, rs, ms), left + 1)

            if cur_node >= len(self.nodes):
                self.nodes.extend(KDTree.leaf(0, [])
                                  for _ in range(cur_node + 1 - len(self.nodes)))

            self.nodes[cur_node].num_parts = 0
            self.nodes[cur_node].split_dim = split_dim
            self.nodes[cur_node].split_val = split_val
            self.nodes[cur_node].m = m
            self.nodes[cur_node].cm = cm
            self.nodes[cur_node].size = size
            self.nodes[cur_node].left = cur_node+1
            self.nodes[cur_node].right = left+1

            return right


def accel_recur(cur_node: int, p: int, particles: Particles, nodes: list[KDTree]) -> npt.NDArray[np.float64]:
    ps, vs, rs, ms = particles
    del particles

    # println!("accel {}", cur_node)
    if nodes[cur_node].num_parts > 0:
        acc = np.zeros(3, dtype=np.float64)
        # print(nodes[cur_node].particles)

        # acc = calc_pp_accel(particles[p],
        #                     particles[nodes[cur_node].particles])
        for i in range(nodes[cur_node].num_parts):
            if nodes[cur_node].particles[i] != p:
                idx = nodes[cur_node].particles[i]
                acc += calc_pp_accel(ps[p],
                                     ps[idx], ms[idx])

        return acc
    else:
        dp = ps[p] - nodes[cur_node].cm
        dist_sqr = dp @ dp
        # println!("dist = {}, size = {}", dist, nodes[cur_node].size)
        if nodes[cur_node].size * nodes[cur_node].size < THETA * THETA * dist_sqr:
            dist = sqrt(dist_sqr)
            magi = -nodes[cur_node].m / (dist_sqr*dist)
            return dp * magi
        else:
            return accel_recur(nodes[cur_node].left, p, (ps, vs, rs, ms), nodes) \
                + accel_recur(nodes[cur_node].right, p,
                              (ps, vs, rs, ms), nodes)


def calc_accel(p: int, particles: Particles, nodes: list[KDTree]):
    return accel_recur(0, p, particles, nodes)


def simple_sim(bodies: Particles, dt: float, steps: int, print_steps: bool = False) -> None:
    p, v, r, m = bodies
    del bodies

    acc = np.zeros((len(p), 3), dtype=np.float64)

    sys = System(len(p))

    dv = np.zeros((len(p), 3), dtype=np.float64)

    for step in range(steps):
        if print_steps:
            print(step)
        sys.indices = [i for i in range(len(p))]

        sys.build_tree(0, len(p), (p, v, r, m), 0)
        if step % 10 == 0:
            print_tree(step, sys.nodes, (p, v, r, m))

        for i in range(len(p)):
            acc[i] = calc_accel(i, (p, v, r, m), sys.nodes)

        acc[:] *= dt  # dt * acc

        v[:] += acc
        np.multiply(dt, v, out=dv)
        p[:] += dv  # dt * v


def print_tree(step: int, tree: list[KDTree], particles: Particles) -> None:
    ps, vs, rs, ms = particles

    with open(f"tree{step}.txt", 'w') as file:
        file.write(f'{len(ps)}\n')
        for n in tree:
            if n.num_parts > 0:
                file.write(f"L {n.num_parts}\n")
                for i in range(n.num_parts):
                    p = n.particles[i]
                    file.write(
                        f"{ps[p][0]} {ps[p][1]} {ps[p][2]}\n")
            else:
                file.write(
                    f"I {n.split_dim} {n.split_val} {n.left} {n.right}\n")


def recur_test_tree_struct(
    node: int,
    nodes: list[KDTree],
    particles: Particles,
    min: tuple[float, float, float],
    max: tuple[float, float, float],
):
    ps, vs, rs, ms = particles

    if nodes[node].num_parts > 0:
        for index in range(nodes[node].num_parts):
            i = nodes[node].particles[index]
            for dim in range(2):
                assert ps[i][dim] >= min[dim], "Particle dim {} is below min. i={} p={} min={}".format(
                    dim,
                    i,
                    ps[i][dim],
                    min[dim])

                assert ps[i][dim] < max[dim], "Particle dim {} is above max. i={} p={} max={}".format(
                    dim,
                    i,
                    ps[i][dim],
                    max[dim])

    else:
        split_dim = nodes[node].split_dim
        tmin = min[split_dim]
        tmax = max[split_dim]
        max[split_dim] = nodes[node].split_val
        recur_test_tree_struct(nodes[node].left, nodes, particles, min, max)
        max[split_dim] = tmax
        min[split_dim] = nodes[node].split_val
        recur_test_tree_struct(nodes[node].right, nodes, particles, min, max)
        min[split_dim] = tmin

File: Parallel/CVersion/src/kdtree.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#include "kdtree.h"
#include "particle.h"

#define MIN(a, b) (((a) < (b)) ? (a) : (b))
#define MAX(a, b) (((a) > (b)) ? (a) : (b))

const size_t NEGS[MAX_PARTS] = {0, 0, 0, 0,
                                0, 0, 0}; // TODO - give maximum value here

KDTree_array_t allocate_node_vec(size_t num_parts) {
  size_t num_nodes = 2 * (num_parts / (MAX_PARTS - 1) + 1);
  KDTree_array_t ret = new_kdtree_array_t(num_nodes);
  return ret;
}

// Returns the index of the last Node used in the construction.
size_t build_tree(size_t_array_t *indices, size_t start, size_t end,
                  const Particle_array_t *particles, size_t cur_node,
                  KDTree_array_t *nodes) {
  size_t np = end - start;
  if (np <= MAX_PARTS) {
    if (cur_node >= nodes->size) {
      KDTree_resize(nodes, cur_node + 1);
    }
    nodes->ptr[cur_node].num_parts = np;
    for (size_t i = 0; i < np; ++i) {
      nodes->ptr[cur_node].particles[i] = indices->ptr[start + i];
    }
    return cur_node;
  } else {
    // Pick split dim and value
    double min_arr[3] = {1e100, 1e100, 1e100};
    double max_arr[3] = {-1e100, -1e100, -1e100};
    double m = 0.0;
    double cm[3] = {0.0, 0.0, 0.0};
    for (size_t i = start; i < end; ++i) {
      m += particles->ptr[indices->ptr[i]].m;
      cm[0] += particles->ptr[indices->ptr[i]].m *
               particles->ptr[indices->ptr[i]].p[0];
      cm[1] += particles->ptr[indices->ptr[i]].m *
               particles->ptr[indices->ptr[i]].p[1];
      cm[2] += particles->ptr[indices->ptr[i]].m *
               particles->ptr[indices->ptr[i]].p[2];

      min_arr[0] = MIN(min_arr[0], particles->ptr[indices->ptr[i]].p[0]);
      min_arr[1] = MIN(min_arr[1], particles->ptr[indices->ptr[i]].p[1]);
      min_arr[2] = MIN(min_arr[2], particles->ptr[indices->ptr[i]].p[2]);
      max_arr[0] = MAX(max_arr[0], particles->ptr[indices->ptr[i]].p[0]);
      max_arr[1] = MAX(max_arr[1], particles->ptr[indices->ptr[i]].p[1]);
      max_arr[2] = MAX(max_arr[2], particles->ptr[indices->ptr[i]].p[2]);
    }
    cm[0] /= m;
    cm[1] /= m;
    cm[2] /= m;
    size_t split_dim = 0;
    if (max_arr[1] - min_arr[1] > max_arr[split_dim] - min_arr[split_dim]) {
      split_dim = 1;
    }
    if (max_arr[2] - min_arr[2] > max_arr[split_dim] - min_arr[split_dim]) {
      split_dim = 2;
    }
    double size = max_arr[split_dim] - min_arr[split_dim];

    // Partition particles on split_dim
    size_t mid = (start + end) / 2;
    size_t s = start;
    size_t e = end;
    while (s + 1 < e) {
      double r;
      do {
        r = (double)rand() / RAND_MAX;
      } while (r ==
               1); // there is a tiny chance r = 1, but we want the range
                   // [s,e) (excluding e), and if r = 1 we would produce pivot=e

      size_t pivot = (size_t)(r * (e - s)) + s;

      size_t c = indices->ptr[s];
      indices->ptr[s] = indices->ptr[pivot];
      indices->ptr[pivot] = c;

      size_t low = s + 1;
      size_t high = e - 1;
      while (low <= high) {
        if (particles->ptr[indices->ptr[low]].p[split_dim] <
            particles->ptr[indices->ptr[s]].p[split_dim]) {
          low += 1;
        } else {
          size_t c = indices->ptr[low];
          indices->ptr[low] = indices->ptr[high];
          indices->ptr[high] = c;
          high -= 1;
        }
      }

      size_t c2 = indices->ptr[s];
      indices->ptr[s] = indices->ptr[high];
      indices->ptr[high] = c2;

      if (high < mid) {
        s = high + 1;
      } else if (high > mid) {
        e = high;
      } else {
        s = e;
      }
    }
    double split_val = particles->ptr[indices->ptr[mid]].p[split_dim];

    // Recurse on children and build this node.
    size_t left =
        build_tree(indices, start, mid, particles, cur_node + 1, nodes);
    size_t right = build_tree(indices, mid, end, particles, left + 1, nodes);

    if (cur_node >= nodes->size) {
      KDTree_resize(nodes, cur_node + 1);
    }
    nodes->ptr[cur_node].num_parts = 0;
    nodes->ptr[cur_node].split_dim = split_dim;
    nodes->ptr[cur_node].split_val = split_val;
    nodes->ptr[cur_node].m = m;
    nodes->ptr[cur_node].cm[0] = cm[0];
    nodes->ptr[cur_node].cm[1] = cm[1];
    nodes->ptr[cur_node].cm[2] = cm[2];
    nodes->ptr[cur_node].size = size;
    nodes->ptr[cur_node].left = cur_node + 1;
    nodes->ptr[cur_node].right = left + 1;

    return right;
  }
}

void calc_pp_accel(const Particle *pi, const Particle *pj, double acc[3]) {
  double dp[3] = {pi->p[0] - pj->p[0], pi->p[1] - pj->p[1],
                  pi->p[2] - pj->p[2]};
  double dist = sqrt(dp[0] * dp[0] + dp[1] * dp[1] + dp[2] * dp[2]);
  double magi = -pj->m / (dist * dist * dist);
  acc[0] += dp[0] * magi;
  acc[1] += dp[1] * magi;
  acc[2] += dp[2] * magi;
}

void accel_recur(size_t cur_node, size_t p, const Particle_array_t *particles,
                 const KDTree_array_t *nodes, double acc[3]) {
  // println!("accel {}", cur_node);
  if (nodes->ptr[cur_node].num_parts > 0) {
    for (size_t i = 0; i < nodes->ptr[cur_node].num_parts; ++i) {
      if (nodes->ptr[cur_node].particles[i] != p) {
        calc_pp_accel(particles->ptr + p,
                      particles->ptr + nodes->ptr[cur_node].particles[i], acc);
      }
    }
  } else {
    double dp[3];
    dp[0] = particles->ptr[p].p[0] - nodes->ptr[cur_node].cm[0];
    dp[1] = particles->ptr[p].p[1] - nodes->ptr[cur_node].cm[1];
    dp[2] = particles->ptr[p].p[2] - nodes->ptr[cur_node].cm[2];
    double dist_sqr = dp[0] * dp[0] + dp[1] * dp[1] + dp[2] * dp[2];
    // println!("dist = {}, size = {}", dist, nodes[cur_node].size);
    if (nodes->ptr[cur_node].size * nodes->ptr[cur_node].size <
        THETA * THETA * dist_sqr) {
      double dist = sqrt(dist_sqr);
      double magi = -nodes->ptr[cur_node].m / (dist_sqr * dist);
      acc[0] += dp[0] * magi;
      acc[1] += dp[1] * magi;
      acc[2] += dp[2] * magi;
    } else {
      accel_recur(nodes->ptr[cur_node].left, p, particles, nodes, acc);
      accel_recur(nodes->ptr[cur_node].right, p, particles, nodes, acc);
    }
  }
}

void calc_accel(size_t p, const Particle_array_t *particles,
                const KDTree_array_t *nodes, double acc[3]) {
  accel_recur(0, p, particles, nodes, acc);
}

void print_tree(int step, const KDTree_array_t *tree,
                const Particle_array_t *particles) {

  if (step > 9999999) {
    fprintf(stderr, "Step too big!\n");
    exit(EXIT_FAILURE);
  }
  char name[8 + 7];
  snprintf(name, 8 + 7, "tree%d.txt", step);

  FILE *file = fopen(name, "w");
  if (file == NULL) {
    fprintf(stderr, "File couldn't be opened!\n");
    exit(EXIT_FAILURE);
  }

  fprintf(file, "%lu\n", particles->size);

  for (size_t i_iter = 0; i_iter < tree->size; ++i_iter) {
    KDTree *n = tree->ptr + i_iter;
    if (n->num_parts > 0) {
      fprintf(file, "L %lu\n", n->num_parts);

      for (size_t i = 0; i < n->num_parts; ++i) {
        size_t p = n->particles[i];
        fprintf(file, "%f %f %f\n", particles->ptr[p].p[0],
                particles->ptr[p].p[1], particles->ptr[p].p[2]);
      }
    } else {
      fprintf(file, "I %lu %f %lu %lu\n", n->split_dim, n->split_val, n->left,
              n->right);
    }
  }

  fclose(file);
}

void simple_sim(Particle_array_t *bodies, double dt, int steps) {
  vect3_array_t acc = new_vect3_array_t(bodies->size);
  for (size_t i = 0; i < bodies->size; ++i) {
    vect3 a = {0.0, 0.0, 0.0};
    acc.ptr[i] = a;
  }
  KDTree_array_t tree = allocate_node_vec(bodies->size);
  size_t_array_t indices = new_range(0, bodies->size);

  for (int step = 0; step < steps; ++step) {
    for (size_t i = 0; i < bodies->size; ++i) {
      indices.ptr[i] = i;
    }

    build_tree(&indices, 0, bodies->size, bodies, 0, &tree);
    if (step % 10 == 0) {
      print_tree(step, &tree, bodies);
    }
    for (size_t i = 0; i < bodies->size; ++i) {
      calc_accel(i, bodies, &tree, acc.ptr[i].v);
    }
    for (size_t i = 0; i < bodies->size; ++i) {
      bodies->ptr[i].v[0] += dt * acc.ptr[i].v[0];
      bodies->ptr[i].v[1] += dt * acc.ptr[i].v[1];
      bodies->ptr[i].v[2] += dt * acc.ptr[i].v[2];
      bodies->ptr[i].p[0] += dt * bodies->ptr[i].v[0];
      bodies->ptr[i].p[1] += dt * bodies->ptr[i].v[1];
      bodies->ptr[i].p[2] += dt * bodies->ptr[i].v[2];
      acc.ptr[i].v[0] = 0.0;
      acc.ptr[i].v[1] = 0.0;
      acc.ptr[i].v[2] = 0.0;
    }
  }

  FREE_ARRAY(acc);
  FREE_ARRAY(tree);
  FREE_ARRAY(indices);
}

vect3_array_t new_vect3_array_t(size_t elem_count) {
  vect3_array_t a = {elem_count, (vect3 *)calloc(elem_count, sizeof(vect3))};

  if (a.ptr == NULL) {
    fprintf(stderr, "calloc failed: it returned NULL\n");
    exit(EXIT_FAILURE);
  }

  return a;
}

KDTree_array_t new_kdtree_array_t(size_t elem_count) {
  KDTree_array_t a = {elem_count, (KDTree *)calloc(elem_count, sizeof(KDTree))};

  if (a.ptr == NULL) {
    fprintf(stderr, "calloc failed: it returned NULL\n");
    exit(EXIT_FAILURE);
  }

  return a;
}

void KDTree_resize(KDTree_array_t *arr, size_t new_elem_count) {
  if (new_elem_count == arr->size) {
    // do nothing
  } else if (new_elem_count > arr->size) {
    arr->ptr = (KDTree *)realloc(arr->ptr, sizeof(KDTree) * new_elem_count);
    arr->size = new_elem_count;
  } else {
    // new_elem_count < arr.size
    arr->ptr = (KDTree *)realloc(arr->ptr, sizeof(KDTree) * new_elem_count);
    arr->size = new_elem_count;
  }
}
